client.go

```diff
@@ -23,9 +35,43 @@ func (c client) doRequest(ctx context.Context, in *request, out interface{}) err
 	if err != nil {
 		return err
 	}
-	bytes, err := c.get(ctx, url)
+
+	resp, err := c.get(ctx, url)
 	if err != nil {
 		return err
 	}
-	return json.Unmarshal(bytes, out)
+	defer closeSilently(resp.Body)
+
+	// OSRM returns both codes 200 and 400 in a case with a body.
+	// In other cases, it returns an unexpected error without a body.
+	// http://project-osrm.org/docs/v5.5.1/api/#requests
+	if resp.StatusCode != http.StatusOK && resp.StatusCode != http.StatusBadRequest {
+		return fmt.Errorf("unexpected http status code %d", resp.StatusCode)
```

Comment:
Maybe read body before checking status code - and include body into error message too (my concern is that body could easily contain some useful information from balancer or smth)

Reply:
updated, PTAL